engine/src/main/java/org/hibernate/validator/internal/engine/validationcontext/AbstractValidationContext.java

```diff
@@ -369,12 +370,13 @@ private boolean isAlreadyValidatedForCurrentGroup(Object value, Class<?> group)
 
 	private void markCurrentBeanAsProcessedForCurrentPath(Object bean, PathImpl path) {
 		// HV-1031 The path object is mutated as we traverse the object tree, hence copy it before saving it
-		Map<Object, Set<PathImpl>> processedPathsPerBean = getInitializedProcessedPathsPerBean();
+		HashMap<ProcessedBean, ArrayList<PathImpl>> processedPathsPerBean = getInitializedProcessedPathsPerBean();
 
-		Set<PathImpl> processedPaths = processedPathsPerBean.get( bean );
+		ProcessedBean processedBean = new ProcessedBean( bean );
+		ArrayList<PathImpl> processedPaths = processedPathsPerBean.get( processedBean );
 		if ( processedPaths == null ) {
-			processedPaths = new HashSet<>();
-			processedPathsPerBean.put( bean, processedPaths );
+			processedPaths = new ArrayList<>();
+			processedPathsPerBean.put( processedBean, processedPaths );
 		}
 
 		processedPaths.add( PathImpl.createCopy( path ) );
```

Comment:
While debugging `ProcessedBeansTrackingCycles1Test`, I see that `ValidatorImpl#validateConstraintsForDefaultGroup`  executes this code (in `#markCurrentBeanAsProcessedForCurrentPath`) twice for the same `Parent` bean ; first time for `Parent.class`, and second time for `Object.class`. Both times `#markCurrentBeanAsProcessedForCurrentPath` is called with the same `Parent` bean and the same `PathImpl` object as arguments.

Since `processedPaths` is an `ArrayList<PathImpl>`, it contains 2 copies of the same `PathImpl`. Was this intended?

I see that  `processedPaths` was changed from a `Set<PathImpl>` to an `ArrayList<PathImpl>`, so before this change, (I think) there would have been just one `PathImpl` in `Set`.

I don't know if this is a real problem; it's just something I noticed.

Reply:
Hmmm. That's unfortunate. It was part of another experiment and shouldn't have been committed with this one.

How do you want us to proceed? Should I fix things and force push? Or we fix it later? Basically, everything in this class should be reverted, except for the change of meaning of the boolean.